trl/trainer/iterative_trainer.py

```diff
@@ -0,0 +1,232 @@
+from accelerate.utils import ProjectConfiguration
+from datasets import Dataset
+from torch.optim import Adam
+from torch.utils.data import DataLoader
+from transformers import PreTrainedModel, PreTrainedTokenizer, PreTrainedTokenizerBase, PreTrainedTokenizerFast
+
+from ..core import PPODecorators, set_seed
+from . import IterativeConfig
+
+
+class IterativeTrainer:
+    """
+    The IterativeTrainer can be used to finetune models with methods that requires some steps between optimization.
+
+    Attributes:
+        **config** (`IterativeConfig`) -- Configuration object for IterativeTrainer.
+        **model** (`PreTrainedModel`) -- Model to be optimized, Hugging Face transformer model with a causal language modeling head.
+            Check the documentation of `PreTrainedModelWrapper` for more details.
+        **tokenizer** (`PreTrainedTokenizerBase`) -- Tokenizer to be used for encoding the
+            data. Check the documentation of `transformers.PreTrainedTokenizer` and
+            `transformers.PreTrainedTokenizerFast` for more details.
+        **optimizer** (`torch.optim.Optimizer`, *optional*) -- Optimizer to be used for training. If no optimizer is
+            provided, the trainer will create an Adam optimizer with the learning rate specified in the configuration
+            object.
+        **data_collator** (DataCollatorForLanguageModeling, *optional*) -- Data collator to be used for training and
+            passed along the dataloader.
+    """
+
+    def __init__(
+        self,
+        config: IterativeConfig = None,
+        model: PreTrainedModel = None,
+        tokenizer: PreTrainedTokenizerBase = None,
+        optimizer: Optional[torch.optim.Optimizer] = None,
+        data_collator: Callable = None,
+    ):
+        """
+        Initialize IterativeTrainer.
+
+        Args:
+            config (`IterativeConfig`):
+                Configuration object for IterativeTrainer.
+            model (`PreTrainedModel`):
+                Hugging Face transformer model.
+            tokenizer (`transformers.PreTrainedTokenizerBase`):
+                Hugging Face tokenizer
+            optimizer (Optional[`torch.optim.Optimizer`]):
+                Optimizer used for training. If `None`, the `Adam` is used as default.
+            data_collator (Optional[function]):
+                Data collator function.
+        """
+
+        self.config = config
+
+        # initial seed for reproducible experiments
+        set_seed(config.seed)
+
+        # Step 0: check positional arguments validity
+        if not isinstance(config, IterativeConfig):
+            raise ValueError(f"config must be a IterativeConfig, got {type(config)}")
+        if not isinstance(tokenizer, (PreTrainedTokenizerBase)):
+            raise ValueError(
+                f"tokenizer must be a PreTrainedTokenizerBase like a PreTrainedTokenizer or a PreTrainedTokenizerFast, got {type(tokenizer)}"
+            )
+
+        # Step 1: Initialize Accelerator
```

Comment:
You can also add a stronger check, one can check if `model.can_generate()` and pass a warning if that method returns `False`